jsk_pcl_ros/src/kinfu_nodelet.cpp

```diff
@@ -33,236 +32,547 @@
-                                      child_frame_id_,
-                                      depth_image->header.stamp,
-                                      ros::Duration(0.1));
-        Eigen::Affine3f odom_camera;
-        tf::transformTFToEigen(trans, odom_camera);
-        if (std::string("")!=parent_frame_id_) // if parent_frame_id is set, child_frame_id must be the top of the tf tree.
+
+    // save texture
+    if (integrate_color_ && (frame_idx_ % pcl::device::kinfuLS::SNAPSHOT_RATE == 1))
+    {
+      boost::mutex::scoped_lock lock(mutex_);
+
+      if (frame_idx_ == 1)
+      {
+        cv::imwrite(save_dir_ + "/occluded.jpg",
+                    cv::Mat::zeros(caminfo_msg->height, caminfo_msg->width, CV_8UC3));
+      }
+
+      cv::Mat image = cv_bridge::toCvShare(color_msg, color_msg->encoding)->image;
+      if (color_msg->encoding == enc::RGB8)
+      {
+        cv::cvtColor(image, image, cv::COLOR_RGB2BGR);
+      }
+      std::stringstream ss;
+      ss << save_dir_ << "/" << ros::Time::now().toNSec() << ".jpg";
+      std::string filename = ss.str();
+      cv::imwrite(filename, image);
+      NODELET_INFO("Saved texture snapshot: %s", filename.c_str());
+
+      pcl::TextureMapping<pcl::PointXYZ>::Camera camera;
+      camera.pose = kinfu_->getCameraPose();
+      camera.focal_length = caminfo_msg->K[0];  // must be equal to caminfo_msg->K[4]
+      camera.height = caminfo_msg->height;
+      camera.width = caminfo_msg->width;
+      camera.texture_file = filename;
+      cameras_.push_back(camera);
+    }
+
+    // publish kinfu origin and slam
+    {
+      Eigen::Affine3f camera_pose = kinfu_->getCameraPose();
+      geometry_msgs::PoseStamped camera_pose_msg;
+      tf::poseEigenToMsg(camera_pose, camera_pose_msg.pose);
+      camera_pose_msg.header.stamp = caminfo_msg->header.stamp;
+      camera_pose_msg.header.frame_id = "kinfu_origin";
+      pub_camera_pose_.publish(camera_pose_msg);
+
+      Eigen::Affine3f camera_to_kinfu_origin = camera_pose.inverse();
+      tf::Transform tf_camera_to_kinfu_origin;
+      tf::transformEigenToTF(camera_to_kinfu_origin, tf_camera_to_kinfu_origin);
+      tf_camera_to_kinfu_origin.setRotation(tf_camera_to_kinfu_origin.getRotation().normalized());
+      tf_broadcaster_.sendTransform(
+        tf::StampedTransform(tf_camera_to_kinfu_origin, caminfo_msg->header.stamp,
+                             caminfo_msg->header.frame_id, "kinfu_origin"));
+
+      if (slam_)
+      {
+        // use kinfu as slam, and publishes tf: map -> fixed_frame_id_ (usually odom_init)
+        try
+        {
+          tf::StampedTransform tf_odom_to_camera;
+          tf_listener_->lookupTransform(
+            fixed_frame_id_, caminfo_msg->header.frame_id, ros::Time(0), tf_odom_to_camera);
+          Eigen::Affine3f odom_to_camera;
+          tf::transformTFToEigen(tf_odom_to_camera, odom_to_camera);
+
+          Eigen::AngleAxisf rotate90_x = Eigen::AngleAxisf(1.57, Eigen::Vector3f::UnitX());
```

Comment:
It's strange that map is upside-down because that means odom_init recorded in rosbag is also upside down. Watching odom_init frame in rosbag will give hint.